### PR TITLE
Improve calculation of gene length

### PR DESCRIPTION
**Pull request by oplantalech**
Following #1583, the calculation of gene length has been revised. Currently, a Perl command is necessary to parse the Gencode file (gtf) into a bed file, while the proposed code directly parses the Gencode file. Regarding the method itself, the proposed code uses Ensembl IDs to differentiate between different genes, which is more accurate than using the symbols (example #1160, where there are two different genes called SNORA31 in the same chromosome; in the Gencode file they have different Ensembl IDs) . 

If different Ensembl IDs point to the same symbol, the method only updates the length of the gene if the chromosome given in the gtf file matches the cytoband from the gene in the database. In the case multiple Ensembl IDs with the same symbol are on the same chromosome (like SNORA31 in #1160) or no cytoband information is available, the length of the last Ensembl ID is taken. Finally, a test for the method has been added, and also the code documentation has been improved. @pieterlukasse @sandertan

### CODE REVIEW

**Comment by oplantalech** [comment]
Also, another improvement in the length calculation is that the end and start bases from the Gencode file are part of the sequence. For example, if for one gene the start base is 2 and the end is 10, the length of the gene is not 8, but 9.

**Comment by pieterlukasse** [comment]
please also add the respective documentation on how to use the new script

**Comment by oplantalech** [comment]
@pieterlukasse documentation added

**Comment by pieterlukasse** [comment]
@oplantalech code looks good. Just some minor code documentation missing.

**Comment by pieterlukasse** [comment]
I did some manual validation using `gencode.v24.annotation.gtf` and the new seed data that has gene lengths based on this fix and this looks good :+1:

**Comment by Sjoerd-van-Hagen** [comment]
@jjgao 

Hi JJ,

This PR has been open for a long time now. Could you make sure it gets reviewed and merged?

Best,

Sjoerd.

**Comment by pieterlukasse** [comment]
@Sjoerd-van-Hagen @jjgao I had reviewed it, but by mistake never sent the review comments somehow. Fortunately github stored them and they are now sent to @oplantalech

**Comment by oplantalech** [comment]
@pieterlukasse I've included your suggestions in the new commit. Can you check again if all you requested has been taken into account?

File: core/src/main/java/org/mskcc/cbio/portal/scripts/ImportGeneData.java [hunk 1]
```diff
@@ -36,95 +36,133 @@
...
-                    }
-                    if (!strAliases.equals("-")) {
-                        aliases.addAll(Arrays.asList(strAliases.split("\\|")));
-                    }
-                    
-                    if (geneSymbol.startsWith("MIR") && type.equalsIgnoreCase("miscRNA")) {
-                        line = buf.readLine();
-                        continue; // ignore miRNA; process seperately
+                String parts[] = line.split("\t");
+                int taxonimy = Integer.parseInt(parts[0]);
```
Comment (n1zea144) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/ImportGeneData.java]: This should be spelled "taxonomy"

File: core/src/main/java/org/mskcc/cbio/portal/scripts/ImportGeneData.java [hunk 2]
```diff
@@ -137,34 +175,49 @@ public static void importData(File geneFile) throws IOException, DaoException {
...
+       
     }
-    
-    private static void logDuplicateGeneSymbolWarning(String symbol, Set<CanonicalGene> genes) {
+
+    private static void logDuplicateGeneSymbolWarning(String symbol, Set<CanonicalGene> genes, boolean isOfficialSymbol) {
         StringBuilder sb = new StringBuilder();
-        sb.append("More than 1 gene has the same symbol ")
-                .append(symbol)
+        if (isOfficialSymbol)
```
Comment (n1zea144) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/ImportGeneData.java]: For consistency in style, I would put braces around this statement.

File: core/src/main/java/org/mskcc/cbio/portal/scripts/ImportGeneData.java [hunk 3]
```diff
@@ -222,6 +397,8 @@ private static int calculateGeneLength(List<long[]> loci) {
                 max = l[1];
             }
         }
+        if (max < min)
```
Comment (n1zea144) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/ImportGeneData.java]: For style consistency I would add braces here.

**Comment by pieterlukasse** [comment]
approved changes based on assumption changes were made. Hard to check since the whole PR has been squashed again...

**Comment by pieterlukasse** [comment]
Were minor comments anyway

**Comment by jjgao** [comment]
@oplantalech: there are some conflicts of this pull request with the current rc branch. Could you please create a pull request to rc with the same change set? Thanks.

**Comment by oplantalech** [comment]
@jjgao The new PR based on RC is here: #2149

**Pull request merged**